core/src/Kokkos_Tuners.hpp

```diff
@@ -0,0 +1,473 @@
+#include <iostream>
+
+namespace Kokkos {
+namespace Tools {
+
+namespace Experimental {
+
+// forward declarations
+SetOrRange make_candidate_set(size_t size, int64_t* data);
+size_t declare_output_type(const std::string&,
+                           Kokkos::Tools::Experimental::VariableInfo);
+void request_output_values(size_t, size_t,
+                           Kokkos::Tools::Experimental::VariableValue*);
+VariableValue make_variable_value(size_t, int64_t);
+VariableValue make_variable_value(size_t, double);
+SetOrRange make_candidate_range(double lower, double upper, double step,
+                                bool openLower, bool openUpper);
+size_t get_new_context_id();
+void begin_context(size_t context_id);
+void end_context(size_t context_id);
+namespace Impl {
+
+/** We're going to take in search space descriptions
+ * as nested maps, which aren't efficient to
+ * iterate across by index. These are very similar
+ * to nested maps, but better for index-based lookup
+ */
+template <typename ValueType, typename ContainedType>
+struct ValueHierarchyNode;
+
+template <typename ValueType, typename ContainedType>
+struct ValueHierarchyNode {
+  std::vector<ValueType> root_values;
+  std::vector<ContainedType> sub_values;
+  void add_root_value(const ValueType& in) noexcept {
+    root_values.push_back(in);
+  }
+  void add_sub_container(const ContainedType& in) { sub_values.push_back(in); }
+  const ValueType& get_root_value(const size_t index) const {
+    return root_values[index];
+  }
+  const ContainedType& get_sub_value(const size_t index) const {
+    return sub_values[index];
+  }
+};
+
+template <typename ValueType>
+struct ValueHierarchyNode<ValueType, void> {
+  std::vector<ValueType> root_values;
+  void add_root_value(const ValueType& in) noexcept {
+    root_values.push_back(in);
+  }
+  const ValueType& get_root_value(const size_t index) const {
+    return root_values[index];
+  }
+};
+
+/** For a given nested map type, we need a way to
+ * declare the equivalent ValueHierarchyNode
+ * structure
+ */
+
+template <class NestedMap>
+struct MapTypeConverter;
+
+// Vectors are our lowest-level, no nested values
+template <class T>
+struct MapTypeConverter<std::vector<T>> {
+  using type = ValueHierarchyNode<T, void>;
+};
+
+// Maps contain both the "root" types and sub-vectors
+template <class K, class V>
+struct MapTypeConverter<std::map<K, V>> {
+  using type = ValueHierarchyNode<K, typename MapTypeConverter<V>::type>;
+};
+
+/**
+ * We also need to be able to construct a ValueHierarchyNode set from a
+ * map
+ */
+
+template <class NestedMap>
+struct ValueHierarchyConstructor;
+
+// Vectors are our lowest-level, no nested values. Just fill in the fundamental
+// values
+template <class T>
+struct ValueHierarchyConstructor<std::vector<T>> {
+  using return_type = typename MapTypeConverter<std::vector<T>>::type;
+  static return_type build(const std::vector<T>& in) { return return_type{in}; }
```

Comment:
This works because `return_type` == `ValueHierarchyNode<T, void>` is an aggregate that has only a `vector<T>` in it.  Any reason it doesn't have an explicit `explicit` constructor for this?  It would certainly be easier to follow the logic.

I was thinking:
`explicit ValueHierarchyNode(std::vector<V> rv) : root_values(std::move(rv)) {}`

Reply:
The traditional reason, didn't think to do it ;)

Will do